core/server/master/src/main/java/alluxio/underfs/MasterUfsManager.java

```diff
@@ -11,17 +11,86 @@
 
 package alluxio.underfs;
 
+import alluxio.exception.InvalidPathException;
+import alluxio.util.UnderFileSystemUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
  * A class that manages the UFS for master servers.
  */
 @ThreadSafe
 public final class MasterUfsManager extends AbstractUfsManager {
+  private static final Logger LOG = LoggerFactory.getLogger(MasterUfsManager.class);
+
+  // The physical ufs state for all active mounts if not the default
+  private ConcurrentHashMap<String, UnderFileSystem.UfsMode> mPhysicalUfsState =
+      new ConcurrentHashMap<>();
 
   /**
    * Constructs the instance of {@link MasterUfsManager}.
    */
   public MasterUfsManager() {}
 
+  @Override
+  public void removeMount(long mountId) {
+    super.removeMount(mountId);
+
+    // Remove any unused physical paths from map
+    for (String physicalUfs : mPhysicalUfsState.keySet()) {
+      boolean found = false;
+      for (UnderFileSystem ufs : mUnderFileSystemMap.values()) {
+        if (ufs.isPathCovered(physicalUfs)) {
+          found = true;
+          break;
+        }
+      }
```

Comment:
You could do something like filter and map. Also, could this get expensive? Alternatively we could keep more state (ie. UFS -> physical ufses and remove the `isPathCovered` API)

Reply:
i'm on board with replacing the `isPathCovered` with `getPhysicalPaths`, and adding state